vim_onenote.ahk

```diff
@@ -45,65 +49,160 @@ IsLastKey(key)
-    SendInput, {ESC} 
-	ToolTip, OneNote Vim Command Mode Active, 0, 0
-return
+
 
 ;--------------------------------------------------------------------------------
 ; Return to InsertMode
 InsertMode:
     ToolTip
     Suspend, On
+    hotkey, ESC, on
 return
 
 ;--------------------------------------------------------------------------------
+
+;  ctrl + [, ESC enter Normal Mode.
+; Not using ctrl + c, as people may still want to use for copying.
+;^c::
+^[::
+ESC::
+    ; Is not affected by insertmode's suspend
+    suspend, permit
+    gosub NormalMode
+return
+
+; imap workings (eg jj) currently can't be implemented because of how 
+; insert mode works.
+
+
+NormalMode:
+    Suspend, Off
+    ToolTip, OneNote Vim Command Mode Active, 0, 0
+return
+
+
+;--------------------------------------------------------------------------------
 +i::
-    SendInput, {Home}
+    send {Home}
     Gosub InsertMode
 return 
 
 i::Gosub InsertMode
 return 
 
 ;--------------------------------------------------------------------------------
+PrepareClipboard(){
+    ; push clipboard to local variable
+    ClipSaved := ClipboardAll
+    ; Clear clipboard to avoid errors
+    Clipboard :=
+}
+
+Copy(){
+    PrepareClipboard()
+    send ^c
+    ClipWait
+}
+
+Paste(){
+    Send %Clipboard%
+    RestoreClipboard()
+}
+
+RestoreClipboard(){
+    ;restore original clipboard
+    Clipboard := ClipSaved
+    ;ClipWait
+    ClipSaved := ; free memory
+}
+
+GetCursorColumn(){
+    send +{home}
+    Copy()
+    position := strLen(Clipboard)
+    RestoreClipboard()
+    return position
+}
+
+ConvertMotionToFunctionName(letter){
+    ;StringLower, test, letter
+    ;msgbox, %test% 
+    if letter is upper
+        return s%letter%
+    else if letter = 0
+        return z
+    else
+        return letter
+}
+
+SelectMotion(){
+    gosub, InsertMode
+    Input, motion, L1
+    gosub, NormalMode
+    ;if motion = i, a or digit, need to wait. If digit, loop motion that many times. If i, g or w, wait for anotther motion.
+    if IsLastKey(motion){
+        send {end}
+        send +{home}
+        return
+    }
+    MoveFunction := ConvertMotionToFunctionName(motion)
+    ;MsgBox, %MoveFunction%
+    send {shift down}
+    %MoveFunction%()
+    send {shift up}
+}
+
+
+w(){
+    send ^{right}
+}
+w::w()
+b(){
+    send ^{left}
+}
+b::b()
+
 ; vi left and right
+h(){ 
+    send {Left}
+}
+h::h()
 
-h::SendInput, {Left}
-return 
-l::SendInput, {Right}
-return 
+l(){
+    send {Right}
+}
+l::l()
 
-;--------------------------------------------------------------------------------
 ; vi up and down.
 ;  Onenote does some magic that blocks up/down processing. See more @ 
 ; (Onenote 2013) http://www.autohotkey.com/board/topic/74113-down-in-onenote/
 ; (Onenote 2007) http://www.autohotkey.com/board/topic/15307-up-and-down-hotkeys-not-working-for-onenote-2007/
-j::SendInput,^{down}
-return 
-k::SendInput,^{up}
-return 
+j(){
+    send {end}
```

Comment:
No, it moves to the end of the line rather than the start (which is what ctrl + up would do). That's still not great, but it's better. Particularly in that it's expected behavior at least _some_ of the time (when you move up to a line that's shorter).
A better version is in the works/completed. However it doesn't work as a generic motion, which this does.